File: go/p4.go
package main

// 转化为求第k最小数
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m := len(nums1)
	n := len(nums2)
	k := int((m + n) / 2)
	if (m+n)&0x01 == 1 {
		return float64(findKthElement(nums1, m, nums2, n, k+1))
	} else {
		return (float64(findKthElement(nums1, m, nums2, n, k)) + float64(findKthElement(nums1, m, nums2, n, k+1))) / 2
	}
}

func findKthElement(nums1 []int, m int, nums2 []int, n int, k int) int {
	if m > n {
		return findKthElement(nums2, n, nums1, m, k)
	}
	if m == 0 {
		return nums2[k-1]
	}
	if k == 1 {
		return min(nums1[0], nums2[0])
	}

	pa := min(int(k/2), m)
	pb := k - pa
	if nums1[pa-1] < nums2[pb-1] {
		return findKthElement(nums1[pa:m], m-pa, nums2, n, k-pa)
	} else if nums1[pa-1] > nums2[pb-1] {
		return findKthElement(nums1, m, nums2[pb:n], n-pb, k-pb)
	} else {
		return nums1[pa-1]
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	result := findMedianSortedArrays([]int{1, 2, 3, 10}, []int{4, 5, 6})
	println(result)
}

File: python/p4.py
class Solution:
    # @return a float
    # 转化为求第k最小数
    def findMedianSortedArrays(self, a, b):
        m = len(a)
        n = len(b)
        k = (m + n) // 2
        return self.__find_kth_element(a, m, b, n, k + 1) if (m + n) & 0x01 \
            else (self.__find_kth_element(a, m, b, n, k) + self.__find_kth_element(a, m, b, n, k + 1)) / 2.0

    def __find_kth_element(self, a, m, b, n, k):
        if m > n:
            return self.__find_kth_element(b, n, a, m, k)
        if m == 0:
            return b[k - 1]
        if k == 1:
            return min(a[0], b[0])

        pa = min(k // 2, m)
        pb = k - pa
        if a[pa - 1] < b[pb - 1]:
            return self.__find_kth_element(a[pa:m], m - pa, b, n, k - pa)
        elif a[pa - 1] > b[pb - 1]:
            return self.__find_kth_element(a, m, b[pb:n], n - pb, k - pb)
        else:
            return a[pa - 1]

